plugins/field-date/src/field_date.ts

```diff
@@ -0,0 +1,262 @@
+   */
+  CURSOR = 'text';
+
+  /**
+   * Class for a date input field. Derived from the Closure library date
+   * picker.
+   * @param value The initial value of the field. Should be in
+   *    'YYYY-MM-DD' format. Defaults to the current date.
+   * @param validator A function that is called to validate
+   *    changes to the field's value. Takes in a date string & returns a
+   *    validated date string ('YYYY-MM-DD' format), or null to abort the
+   *    change.
+   */
+  constructor(value?: string, validator?: FieldDateValidator) {
+    super(value, validator);
+  }
+
+  /**
+   * Constructs a FieldDate from a JSON arg object.
+   * @param options A JSON object with options (date).
+   * @returns The new field instance.
+   * @package
+   * @nocollapse
+   */
+  static fromJson(options: FieldDateFromJsonConfig): FieldDate {
+    // `this` might be a subclass of FieldDate if that class doesn't
+    // override the static fromJson method.
+    return new this(options.date, undefined);
+  }
+
+  /* eslint-disable @typescript-eslint/naming-convention */
+  /**
+   * Ensures that the input value is a valid date.
+   * @param newValue The input value. Ex: '2023-04-28'
+   * @returns A valid date string, or null if invalid.
+   * @override
+   */
+  protected doClassValidation_(newValue?: string): string | null {
+    if (!newValue) return null;
+
+    const newDate = typeof newValue === 'string' ? new Date(newValue) : null;
+    if (!newDate || isNaN(newDate.getTime())) return null;
+
+    // NOTE: 'newValue' should be a valid date format here.
+    if (isISOFormat(newValue)) return newValue;
+
+    // Assume the time needs to be corrected.
+    return toLocalISOString(newDate);
+  }
+
+  /**
+   * Get the text to display on the block when the input hasn't spawned in.
+   *
+   * @returns The text to display on the block.
+   * @override
+   */
+  protected getText_(): string | null {
+    const value = this.getValue();
+    if (!value) return null;
+    // NOTE: There may be discrepancies between the text and the input based on
+    // browser. For example, 'en-US' would display the text '2/14/2020', then
+    // clicking in Safari on iOS would display 'Feb 14, 2020'.
+    return getLocaleDateString(value);
+  }
+
+  /**
+   * Renders the field. If the picker is shown make sure it has the current
+   * date selected.
+   */
+  protected render_() {
+    super.render_();
+  }
+
+  /**
+   * Shows the inline free-text editor on top of the text along with the date
+   * editor.
+   * @param e Optional mouse event that triggered the field to
+   *     open, or undefined if triggered programmatically.
+   * @param _quietInput Quiet input.
+   * @override
+   */
+  protected showEditor_(e?: Event, _quietInput?: boolean) {
+    if (this.sourceBlock_) {
+      // NOTE: Disable modal inputs to use default browser inputs on mobile.
+      this.sourceBlock_.workspace.options.modalInputs = false;
```

Comment:
I'm not sure that we want to disable modalinputs for the whole workspace to get the date picker to work. I would rather:

1) Pass `quietInput` as `true` to super, but [focus on it anyway](https://github.com/google/blockly/blob/a677355f7319140f68c13c3468470b0ecd9d25b8/core/field_input.ts#L340C5-L345C6).
2) Tell external devs to disable modal inputs in their workspace config if they want to use this plugin. We could also log a warning from here if we run into enabled `modalInputs`.

Reply:
I went ahead and implemented `option 1` of your suggestions. It seemed the least invasive. TBH, I'm not in tune with the use case of needing to set quietInput + avoid focusing on the input, though, so I'm not sure how important that is.